Timeout (#30)
monstar-lab-oss/API-manifesto: README.md

# Timeouts

> Generally APIs should respond in less than 250ms on a wired connection. There needs to be a special reason for exeeding that. And the developer needs to understand that it will be harder / more expensive to scale the backend if response times are high.

Reviewer: ```suggestion
Generally APIs should respond in less than 250ms on a wired connection. There needs to be a special reason for exceeding that. Further, it is important to understand that it will be harder and more expensive to scale the backend if response times are high.
```

> It's common that the webserver configuration will timeout the request in 30 or 60 sec. 

Reviewer: ```suggestion
It's common that the webserver configuration will timeout the request after 30 or 60 seconds. 
```

## Client to server

> Since you never know what network the client is on, if they are in a metro or 100mBit wifi. Response time can very a lot for several reason.

Reviewer 2: ```suggestion
Since you never know what network the client is on, if they are in a metro or 100mBit wifi. Response time can vary a lot for several reason.
```

> Since you never know what network the client is on, if they are in a metro or 100mBit wifi. Response time can vary a lot for several reason.

Reviewer: ```suggestion
Since you never know what network the client is on, if they are in a metro or on a 100 Mbit wifi connection,  response times can vary a lot for several reasons.
```

Therefore set timeouts to:

<details>
<summary>Click to see examples</summary>

### ✅

__Default:__ 15 sec

__File upload APIs:__ Align with web server timeout (eg 30 or 60 sec)

> _If you are going to upload files above 5mb, consider having client upload directly to AWS S3, Dropbox etc. And sending path to server._

Reviewer: ```suggestion
_If you are going to upload files above 5mb, consider having the client upload directly to AWS S3, Dropbox etc. and afterwards sending the file path to the backend._
```

### ⛔️

__+ 30sec__

> If API requests are taking more than 2sec on wired connection, reconsider change of api design. 

Reviewer 3: ```suggestion
If API requests are taking more than 2 sec on a wired connection, reconsider change of API design. 
```

> If API requests are taking more than 2 sec on a wired connection, reconsider change of API design. 

Reviewer: ```suggestion
If API requests are taking more than 2 sec on a wired connection, consider changing the API design. 
```

> Eg: Put the operation in a queue system like SQS, Redis, Beanstalkd and inform the client about operation is complete by push notification, web socket, email etc

Reviewer 3: ```suggestion
Eg: Put the operation in a queue system like SQS, Redis, Beanstalkd and inform the client about operation is complete by push notification, web socket, email etc.
```

</details>

## Server to server

> Server to service APIs should always be very stable due to connection is wired and stable. Therefore we can be much more aggressive about timeouts

Reviewer 3: ```suggestion
Server to service APIs should always be very stable due to connection being wired and stable. Therefore we can be much more aggressive about timeouts.
```

> Server to service APIs should always be very stabile due to connection is wiried and stable. Therefore we can be much more aggresive about timeouts

Reviewer 2: ```suggestion
Server to service APIs should always be very stabile due to connection is wiried and stable. Therefore we can be much more aggresive about timeouts
```
```suggestion
Server to service APIs should always be very stabile due to connection is wiried and stable. Therefore we can be much more aggresive about timeouts
```
```suggestion
Server to service APIs should always be very stable due to connection is wired and stable. Therefore we can be much more aggressive about timeouts
```

> Server to service APIs should always be very stabile due to connection is wiried and stable. Therefore we can be much more aggresive about timeouts

Reviewer 2: ```suggestion
Server to service APIs should always be very stable due to connection is wired and stable. Therefore we can be much more aggressive about timeouts
```

<details>
<summary>Click to see examples</summary>

### ✅

> Depending on service: 1-5sec timeouts. 

Reviewer 3: ```suggestion
Depending on service: 1-5 sec timeouts.
```

> Implement a retry system. If the server is not responding in 1-5 sec, there is a high chance they never respond. Just fail and retry, up to a max og 3-5 retries, and then throw exception

Reviewer 3: ```suggestion
Implement a retry system. If the server is not responding in 1-5 sec, there is a high chance they never respond. Just fail and retry, up to a max of 3-5 retries, and then throw exception.
```

> Implement a retry system. If the server is not responding in 1-5 sec, there is a high chance they never respond. Just fail and retry, up to a max of 3-5 retries, and then throw exception.

Reviewer: ```suggestion
Implementing a retry system is strongly advised. If the server is not responding in 1-5 sec, there is a high chance they never respond. Just fail and retry, up to a max of 3-5 retries, and then throw an exception.
```
